compiler/rustc_lexer/src/cursor.rs

```diff
@@ -91,7 +91,17 @@ impl<'a> Cursor<'a> {
         // It was tried making optimized version of this for eg. line comments, but
         // LLVM can inline all of this and compile it down to fast iteration over bytes.
         while predicate(self.first()) && !self.is_eof() {
-            self.bump();
+            #[cfg(debug_assertions)]
+            {
+                self.bump();
+            }
+
+            #[cfg(not(debug_assertions))]
```

Comment:
I'm not a contributor but I personally wouldn't like that this executes largely different code as that sounds like potential for bad heisenbugs.

Reply:
Yes, I have considered this. But the `bump` method only does something beyond advancing the cursor when asserts are enabled in the first place and (storing the previous byte).

Reply:
> Edit: Did you check how much this part improves performance?

That's why this is a draft PR and I'm going to run perf on that.